### PR TITLE
Username and Password regex added

### PR DESCRIPTION
**Pull request by aksinha-nerdapplabs**
[Issue# 8](https://github.com/nerdapplabs/authOauth/issues/8)

### CODE REVIEW

File: src/ApiBundle/Resources/config/config.yml [hunk 1]
```diff
@@ -3,6 +3,13 @@ parameters:
     # app_locales: en|fr|de|es|cs|nl|ru|uk|ro|pt_BR|pl|it|ja|id|ca|sl
     app_locales: en|fr|hi
 
+    # Password matching expression. Match all alphanumeric character and predefined wild characters.
+    # Password must consists of at least 8 characters and not more than 15 characters.
+    password_regex: '/^([a-zA-Z0-9@*#]{8,15})$/'
```
Comment (praveen-garg) [inline src/ApiBundle/Resources/config/config.yml]: don;t we have this in validation.yml ?

File: src/ApiBundle/Resources/config/config.yml [hunk 2]
```diff
@@ -3,6 +3,13 @@ parameters:
...
+
+    # username at least 3 of those characters, but no more than 16, any lowercase letter (a-z), number (0-9), an underscore, or a hyphen
+    username_regex: '/^[a-z0-9_-]{3,16}$/'
```
Comment (praveen-garg) [inline src/ApiBundle/Resources/config/config.yml]: don;t we have this in validation.yml ?

File: src/ApiBundle/Controller/AuthController.php [hunk 3]
```diff
@@ -503,12 +440,30 @@ public function editProfileAction()
...
+                                    "error_description" => $error->getMessage(),
+                                    "show_message" => $this->get('translator')->trans($constraint->payload['api_error'], array(), 'messages', $request->getLocale())
+                                  );
+                array_push($errorArray, $errorItem);
+            }
+            return new JsonResponse($errorArray);
+          }
 
+        // Everything ok, now update the user record
         $userManager->updateUser($user);
```
Comment (praveen-garg) [inline src/ApiBundle/Controller/AuthController.php]: move         ` $userManager = $this->get('fos_user.user_manager'); ` line on top of this;

File: src/ApiBundle/Controller/AuthController.php [hunk 4]
```diff
@@ -503,12 +440,30 @@ public function editProfileAction()
...
         $data = $request->request->all();
 
-        $this->handleKeyUsername($user, $request);
-        $this->handleKeyEmail($user, $request);
-        $this->handleKeyFirstname($user, $request);
-        $this->handleKeyLastname($user, $request);
-        $this->handleKeyDob($user, $request);
+        $user->setFirstname(array_key_exists('firstname', $data) ? $data['firstname'] : $user->getFirstname() );
+        $user->setLastname(array_key_exists('lastname', $data) ? $data['lastname'] : $user->getLastname() );
+        $user->setDob( array_key_exists('dob', $data) ? $data['dob'] : $user->getDob() );
```
Comment (praveen-garg) [inline src/ApiBundle/Controller/AuthController.php]: for above 3, can we please do:
```
$fName = array_key_exists('firstname', $data) ? $data['firstname'] : $user->getFirstname();
$user->setFirstname ($fName);
```

File: src/ApiBundle/Controller/AuthController.php [hunk 5]
```diff
@@ -503,12 +440,30 @@ public function editProfileAction()
...
+        $errors = $validator->validate($user, null, array('profile_edit'));
+
+        if (count($errors) > 0) {
+            $errorArray = [];
+            foreach ($errors as $error) {
+                $constraint = $error->getConstraint();
+                $errorItem = array(
+                                    "code" => 400,
+                                    "error" =>  "Bad Request",
+                                    "error_description" => $error->getMessage(),
```
Comment (praveen-garg) [inline src/ApiBundle/Controller/AuthController.php]: I believe, it should be 
```
"code" => 400,                                  
"error" =>  "Bad Request",
"error_details" => [ {"description": <>, "show_message": <>},
                       {"description": <>, "show_message": <>},
]

```

**Review by praveen-garg — CHANGES_REQUESTED** [review]

File: src/ApiBundle/Controller/AuthController.php (hunk 5 above)
Comment (praveen-garg) [reply src/ApiBundle/Controller/AuthController.php]: I'm not sure if @mshariq-nerd @ajabble suggest, to keep `error_type` as well? and  in `error_details` we should add 'field_name' = "FIRST_NAME", 'field_name' = "EMAIL" .... as well

File: src/ApiBundle/Resources/translations/messages.en.yml [hunk 6]
```diff
@@ -145,3 +145,4 @@ api:
   show_error_reset_req:  Password reset request already received
   show_error_username_policy:  Username should be 3-16 characters long with any lowercase letter (a-z), number (0-9), an underscore, or a hyphen
   show_error_password_policy:  Password should be 8-15 characters long and must contain alphanumeric and @*# characters
+  show_error_role:  Sorry! Wrong Role!
```
Comment (praveen-garg) [inline src/ApiBundle/Resources/translations/messages.en.yml]: You are not authorised for this? "something like that" makes more sense!

Comment (aksinha-nerdapplabs) [reply src/ApiBundle/Resources/translations/messages.en.yml]: Role almost would never be needed

**Comment by aksinha-nerdapplabs** [comment]
@praveen-garg all done. Pl check.

File: src/ApiBundle/Controller/AuthController.php [hunk 7]
```diff
@@ -248,23 +248,7 @@ public function postRegisterAction()
...
-                $this->logMessage(400, $errorItem['error_description'] );
-            }
-            return new JsonResponse(array(
-                          "code" => 400,
-                          "error" =>  "Bad Request",
-                          "error_description" => $errorArray[0]['error_description'],
-                          "show_message" => $errorArray[0]['show_message'],
-                          'errors' => $errorArray
-                      ));
+           return $this->reportValidationErrors($request, $errors);
```
Comment (praveen-garg) [inline src/ApiBundle/Controller/AuthController.php]: `+           return $this->reportValidationErrors($errors, $request->); `
can we make it,  
`+           return $this->reportValidationErrors($errors, $request->getLocale()); `

**Pull request merged**